### PR TITLE
Timeline interaction fixes

### PR DESCRIPTION
**Pull request by patrickdalla**
Correction to bug of OOM related to non released object references to GC.
Also, list of item ids related to each timeline bar are not kept in memory.
A intermediary index is also persisted to improve navigation response times.

### CODE REVIEW

**Comment by tc-wleite** [comment]
> @tc-wleite, when you have time, could you test this with your huge case?

I tested it here, memory consumption was clearly reduced (visible in the first cache building). 
However, I still got out-of-memory exceptions after ~30 minutes (from case opening start), and couldn't use the case, even after waiting a few hours.

**Comment by patrickdalla** [comment]
I am investigating here. There is a cache of date objects created to reduce
time spent on string date parsing. Is an array list of all existent date
entries in a specific field. As you'said, the multicase is made of many
evidences in fastmode, so almost certainly they are real file system files
with MAC times.
This cache scope is an event timestamp counting acumulation of a specific
time period.
So, for each run this cache can consume 125M*24bytes=2.8gb.
There is an equivalent cache scheme for timeperiod, that is the
corresponding timeperiod that each entry matches. For the second and minute
times periods this can be also very big.
As these countings run concurrently, it seems that these caches can be some
of the OOM causes.
Tomorrow I will try to ignore this cache. Maybe we can improve the time
string parsing.

Em ter., 28 de mar. de 2023 13:25, Wladimir Leite ***@***.***>
escreveu:

**Comment by tc-wleite** [comment]
I just uploaded a Head Dump to our WebDAV server (/Temp/Wladimir folder).

**Comment by lfcnassif** [comment]
> I just uploaded a Head Dump to our WebDAV server (/Temp/Wladimir folder).

Thank you @tc-wleite. I took a look at the dump, some results:
image

image

image

Seems a lot of memory are being used by Integer object instances in CacheEventEntry objects by an ArrayList<Integer> array, about 5GB from your 15GB heap dump. On last commit I replaced it by Eclipse collections IntArrayList class that uses a primitive array internally. Please give it a try and if it is not enough, send me another heap dump so I can take another look.

**Comment by lfcnassif** [comment]
We can also see ItemId instances are using 3.5GB in this huge case, but it is the max usage for this structure. On the other hand, I guess the ArrayList\<Integer\> overhead in CacheEventEntry was growing up, since your heap is ~15GB (not ~30GB) and I think you took it much before OOME while the graph caches were still being populated.

**Comment by tc-wleite** [comment]
> PS: Your case seems bigger than I thought, it has 157M items?

Yes, I am still adding some images using "--append" to it, so it has more items than when I first opened the issue.

**Comment by tc-wleite** [comment]
> On the other hand, I guess the ArrayList<Integer> overhead in CacheEventEntry was growing up, since your heap is ~15GB (not ~30GB) and I think you took it much before OOME while the graph caches were still being populated.

Yes, I took the head dump before the used memory reached ~32 GB, but that happens right after that (less than one minute).

**Comment by tc-wleite** [comment]
> On last commit I replaced it by Eclipse collections IntArrayList class that uses a primitive array internally. Please give it a try and if it is not enough, send me another heap dump so I can take another look.

@lfcnassif, I opened the case using the version with your latest changes, and the memory usage was reduced a lot as expected.
I was enough to open the case and build all timeline caches. 
It took a while (~1 hour) to complete the process (when the CPU usage became low and used memory became stable), but that seems fine to me as this is a very large case. 
It took ~20 minutes to start building the other caches (after the first one was built). Before that, the UI was blocked, but that seems acceptable too, as the cache building will happen only once.

I took a head dump (after ~30 minutes from the start) and uploaded it.
The log, the memory and CPU usage graphs:

**Comment by patrickdalla** [comment]
Great, it seems it worked with the usage of IntArrayList changed by Nassif. 
I just pushed a better time string parser (~10x faster than the previous).

**Comment by lfcnassif** [comment]
Thank you both! I Will look at the new dump when I have time, I'm very busy this week... It was took when heap usage was high, right @tc-wleite?

> It took ~20 minutes to start building the other caches (after the first one was built). Before that, the UI was blocked

I think we should fix this, even if it is blocked just at the first time, it's not good in my opinion.

**Comment by lfcnassif** [comment]
Current heap usage details:
image
image
image

I'll try to replace the eclipse IntArrayList for a SparseBitSet to see how it behaves. Could you test it again @tc-wleite?

**Comment by tc-wleite** [comment]
> I'll try to replace the eclipse IntArrayList for a SparseBitSet to see how it behaves. Could you test it again @tc-wleite?

Sure!

**Comment by lfcnassif** [comment]
Done. Please let me know how behaves the new memory footprint chart, thank you!

File: iped-app/src/main/java/iped/app/timelinegraph/cache/persistance/CachePersistance.java [hunk 1]
```diff
@@ -130,16 +136,19 @@ public void run() {
...
         for (File f : baseDir.listFiles()) {
             if (f.getName().equals(className.getSimpleName())) {
+                newCache.setIndexFile(className.getSimpleName(),baseDir);
                 ArrayList<CacheTimePeriodEntry> times = new ArrayList<CacheTimePeriodEntry>();
                 newCache.put(f.getName(), times);
+                /*
                 if (!loadEventNewCache(times, f)) {
                     throw new IOException("File not committed:" + f.getName());
                 }
+                */
```
Comment (lfcnassif) [inline iped-app/src/main/java/iped/app/timelinegraph/cache/persistance/CachePersistance.java:151]: @patrickdalla did this change disabled cache loading or am I wrong?

Comment (lfcnassif) [reply iped-app/src/main/java/iped/app/timelinegraph/cache/persistance/CachePersistance.java]: I just reverted it.

**Comment by patrickdalla** [comment]
Yes. Maybe the terminology (names of methods and classes) needs be
corrected. This is an index. The real 'window' cache is done by dataset. As
we limited the max window per timeperiod, the response time is enough
without loading all the index (cache).



Em qua., 29 de mar. de 2023 16:34, Luis Filipe Nassif <
***@***.***> escreveu:

**Comment by lfcnassif** [comment]
@patrickdalla I noticed these exceptions in UI log with this PR, they are not thrown on master:
```
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
	at iped.app.timelinegraph.IpedChartPanel.removeFromDatePart(IpedChartPanel.java:590)
	at iped.app.timelinegraph.datasets.IpedTimelineDataset.caseSearchFilterLoad(IpedTimelineDataset.java:374)
	at iped.app.timelinegraph.datasets.IpedTimelineDataset$1.run(IpedTimelineDataset.java:172)
	at java.base/java.lang.Thread.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: The 'year' argument must be in range 1900 to 9999.
	at org.jfree.chart.date.SpreadsheetDate.<init>(SpreadsheetDate.java:85)
	at org.jfree.chart.date.SerialDate.createInstance(SerialDate.java:735)
	at org.jfree.data.time.Day.initUsing(Day.java:170)
	at org.jfree.data.time.Day.<init>(Day.java:162)
	at org.jfree.data.time.Day.<init>(Day.java:130)
	... 11 more
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
	at iped.app.timelinegraph.IpedChartPanel.removeNextFromDatePart(IpedChartPanel.java:578)
	at iped.app.timelinegraph.datasets.IpedTimelineDataset.caseSearchFilterLoad(IpedTimelineDataset.java:375)
	at iped.app.timelinegraph.datasets.IpedTimelineDataset$1.run(IpedTimelineDataset.java:172)
	at java.base/java.lang.Thread.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: The 'year' argument must be in range 1900 to 9999.
	at org.jfree.chart.date.SpreadsheetDate.<init>(SpreadsheetDate.java:85)
	at org.jfree.chart.date.SerialDate.createInstance(SerialDate.java:735)
	at org.jfree.data.time.Day.initUsing(Day.java:170)
	at org.jfree.data.time.Day.<init>(Day.java:162)
	at org.jfree.data.time.Day.<init>(Day.java:130)
	... 11 more
```

**Comment by patrickdalla** [comment]
Yes. I saw them today. It happens because I put the start and end dates of
the first timeline range in a range outside the supported jfreechart range
(1/1/1900 to 9999). It seems not to lead in erroneous result. I can correct
this tomorrow. I think these min/max dates are declared as constants.

Em qua., 29 de mar. de 2023 17:23, Luis Filipe Nassif <
***@***.***> escreveu:

**Comment by tc-wleite** [comment]
> Done. Please let me know how behaves the new memory footprint chart, thank you!

I tried 2 times (the second one with the first cache already built), but the opening process didn't finish (even after ~2 hours).
Somehow, it seems that more memory is used now.

image

**Comment by lfcnassif** [comment]
Thank you @tc-wleite and sorry. I think next week I'll have time to test other SparseBitSet implementations, without bothering you, that exist out there:
- https://lucene.apache.org/core/9_2_0/core/org/apache/lucene/util/SparseFixedBitSet.html
- https://github.com/RoaringBitmap/RoaringBitmap
- https://github.com/lemire/javaewah

**Comment by tc-wleite** [comment]
No problem at all. I can run other tests with this large case where you guys need.

**Comment by lfcnassif** [comment]
Hi @tc-wleite, I run a few tests here with some libraries, the zaxxer SparseBitSet really uses a ton of memory. [RoaringBitmap](https://github.com/RoaringBitmap/RoaringBitmap) was better, but used just 10% less heap than Eclipse primitive IntArrayList in a case with 2M items. Please test with your huge case, maybe the behavior is better (or worse...).

**Comment by patrickdalla** [comment]
good to know. to implement #39 I used sparsebitset. I will try this new one.

I think that to improve, i will have to flush on disk intermediary results
and merge them at the end. But as this can take some time it can be done on
a new PR. This one already saves a lot of mem.

Em qui., 30 de mar. de 2023 15:32, Luis Filipe Nassif <
***@***.***> escreveu:

**Comment by patrickdalla** [comment]
I've made  some tests with this RoaringBitmap. For a lot of Bitset with small number of bits set (=100 bit set per bitset) IntArraList was still better (both in mem usage as in loading time). This kind of bitsets will be very common on small periods (like hour, second and minute).
In test with larger bitsets (=10000 bit set per bitset), mem usage was ~half of the same code with IntArraList, but loading time was 3x worse.

image
image

**Comment by tc-wleite** [comment]
I am testing with the large case. I will report the results later.

**Comment by patrickdalla** [comment]
Maybe RoaringBitmap serialization and deserialization times can overcome this worse compared loading time, as it can save and load the compressed bitmap version, not needing to compress or decompress it again.

**Comment by patrickdalla** [comment]
It would save some disk space too.

**Comment by tc-wleite** [comment]
> Hi @tc-wleite, I run a few tests here with some libraries, the zaxxer SparseBitSet really uses a ton of memory. [RoaringBitmap](https://github.com/RoaringBitmap/RoaringBitmap) was the best between those I tested, but used just 10% less heap than Eclipse primitive IntArrayList in a case with 2M items. Please test with your huge case, maybe the behavior is better (or worse...).

This new version using `RoaringBitmap` seems the best option so far (at least for the huge case).
I "lost" the memory usage graph (it shows only the last hour, and I took too long to check back).
But I watched the critical part, and it didn't reach the same high peak observed with the `IntArrayList`.

The time spent to build the cache was similar (considering the last one), but the others were a bit faster (log below).
Unfortunately I am still adding items to this case (I am using it for something else), so the comparison is not accurate, as there are about 20M more items now. 
So if we consider this growth, this last option seems even better. 

```text
2023-03-30 19:31:11	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class org.jfree.data.time.Year]]: 256596ms
2023-03-30 19:31:17	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class org.jfree.data.time.Quarter]]: 262911ms
2023-03-30 19:31:25	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class org.jfree.data.time.Month]]: 271118ms
2023-03-30 19:31:27	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class org.jfree.data.time.Week]]: 272653ms
2023-03-30 19:39:11	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class org.jfree.data.time.Hour]]: 737001ms
2023-03-30 19:47:32	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class org.jfree.data.time.Second]]: 1237714ms
2023-03-30 19:53:08	[INFO]	[timelinegraph.cache.IndexTimeStampCache]			Time to build time cache of [[class iped.jfextensions.model.Minute]]: 1573986ms
```

**Comment by patrickdalla** [comment]
Very good.
I pushed commit to persists those Roaring bitstreams using their own serialization methods. For the case I tested it saved ~8% disk space only.
I implemented logic to identify if these serialization methods were used to create the real index. If not, it loads as before, keeping compatibility with already created indexes. 
Anyway, as I've suggested, the next step to improve this would be to save intermediary counting results to release memory, and merge them at the end. This specific implementation could be kept to next PR (not closing the issue or creating a new one). What do you say @lfcnassif ?

**Comment by lfcnassif** [comment]
> Anyway, as I've suggested, the next step to improve this would be to save intermediary counting results to release memory, and merge them at the end. This specific implementation could be kept to next PR (not closing the issue or creating a new one). What do you say @lfcnassif ?

Seems great to me! (if the merge function doesn't spend lots of memory)

**Comment by lfcnassif** [comment]
The first zoom operation (using a rectangle selection) and without a pre-built cache, is freezing the UI for a while, it didn't happen before this PR, I took 3 different thread dumps that show UI is freezing here:
```
"AWT-EventQueue-0" #39 prio=6 os_prio=0 cpu=9265.63ms elapsed=62.58s tid=0x0000022aa3828800 nid=0x7908 waiting on condition  [0x00000057725fa000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.13/Native Method)
        - parking to wait for  <0x00000001726842e8> (a java.util.concurrent.Semaphore$NonfairSync)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.13/Unknown Source)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.13/Unknown Source)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(java.base@11.0.13/Unknown Source)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@11.0.13/Unknown Source)
        at java.util.concurrent.Semaphore.acquire(java.base@11.0.13/Unknown Source)
        at iped.app.timelinegraph.datasets.IpedTimelineDataset.waitLoaded(IpedTimelineDataset.java:578)
        at iped.app.timelinegraph.datasets.IpedTimelineDataset.notifyVisibleRange(IpedTimelineDataset.java:1246)
        at iped.app.timelinegraph.IpedDateAxis.forceRange(IpedDateAxis.java:389)
        at iped.app.timelinegraph.IpedDateAxis.setRange(IpedDateAxis.java:410)
        at org.jfree.chart.axis.DateAxis.setRange(DateAxis.java:442)
        at org.jfree.chart.axis.DateAxis.zoomRange(DateAxis.java:1751)
        at org.jfree.chart.plot.XYPlot.zoomDomainAxes(XYPlot.java:4799)
        at iped.app.timelinegraph.IpedChartPanel.zoom(IpedChartPanel.java:951)
        at iped.jfextensions.chart.ChartPanel.mouseReleased(ChartPanel.java:1884)
        at iped.app.timelinegraph.IpedChartPanel.mouseReleased(IpedChartPanel.java:413)
        at java.awt.AWTEventMulticaster.mouseReleased(java.desktop@11.0.13/Unknown Source)
        at java.awt.AWTEventMulticaster.mouseReleased(java.desktop@11.0.13/Unknown Source)
        at java.awt.Component.processMouseEvent(java.desktop@11.0.13/Unknown Source)
        at javax.swing.JComponent.processMouseEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.Component.processEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.Container.processEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.Component.dispatchEventImpl(java.desktop@11.0.13/Unknown Source)
        at java.awt.Container.dispatchEventImpl(java.desktop@11.0.13/Unknown Source)
        at java.awt.Component.dispatchEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.LightweightDispatcher.retargetMouseEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.LightweightDispatcher.processMouseEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.LightweightDispatcher.dispatchEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.Container.dispatchEventImpl(java.desktop@11.0.13/Unknown Source)
        at java.awt.Window.dispatchEventImpl(java.desktop@11.0.13/Unknown Source)
        at java.awt.Component.dispatchEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventQueue$4.run(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventQueue$4.run(java.desktop@11.0.13/Unknown Source)
        at java.security.AccessController.doPrivileged(java.base@11.0.13/Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(java.base@11.0.13/Unknown Source)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(java.base@11.0.13/Unknown Source)
        at java.awt.EventQueue$5.run(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventQueue$5.run(java.desktop@11.0.13/Unknown Source)
        at java.security.AccessController.doPrivileged(java.base@11.0.13/Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(java.base@11.0.13/Unknown Source)
        at java.awt.EventQueue.dispatchEvent(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.13/Unknown Source)
        at java.awt.EventDispatchThread.run(java.desktop@11.0.13/Unknown Source)

```

**Comment by lfcnassif** [comment]
PS: I think we should avoid calling Semaphore.acquire() (or acquiring monitor locks for long tasks) in UI thread, it doesn't seem a good approach to me.

**Comment by patrickdalla** [comment]
A possible idea also can be to run these index creations in separated processes. If any OOM occurs, it would not stop main IPED, only leaving him without timeline chart.

**Comment by patrickdalla** [comment]
I've found some points in code that can be improved but that impact other
modules than timeline, so better to leave them for other PR, line bookmark
filtering via RoaringBitmaps.

For the problem of better memory utilization when navigating, it is
finished and it should be the purpose of this PR.

For better memory utilization on timeline index creation, I think that only
with partial accumulation data persistence for latter index merging. But it
can take some time, and can be let to latter PR too.

Em qua., 12 de abr. de 2023 09:40, Luis Filipe Nassif <
***@***.***> escreveu:

**Comment by lfcnassif** [comment]
Great, thank you @patrickdalla!

**Comment by lfcnassif** [comment]
OOM errors, if not caused by increasing leaks or huge array allocations caused by reading array lengths from untrusted data, strictly speaking are not bugs, but inefficient resource usage that may be optimized. I talked to @patrickdalla privately and he didn't find actual bugs present in previous versions fixed by this. So, I'll change the original issue to enhancement and I'll integrate it into master after testing, but not on 4.1.x version for now.

**Review by lfcnassif — APPROVED** [review]
Thank you @patrickdalla!

**Pull request merged**